pq-crypto/EVP_kem.h

```diff
@@ -0,0 +1,179 @@
+*
+* Description: Uses public key to create cipher text and secrect key.
+*
+* Arguments: pq_kem_params
+* The following fields of pq_kem_params are used for encapsualte:
+* - unsigned char *cipher_text: pointer to output cipher text
+* (already allocated array of bytes)
+* - unsigned char *shared_secret: pointer to output shared secrect
+* (already allocated array of bytes)
+* - const unsigned char *public_key: pointer to input constant public key
+*
+* Returns EVP_PQ_SUCCESS on successful encapsulation,
+* return EVP_PQ_FAILURE otherwise and on error
+**************************************************/
+int EVP_kem_encapsulate(pq_kem_params *kem_params);
+
+
+/*************************************************
+* Name: EVP_kem_dec
+*
+* Description: Generates shared secret for public key and cipher text
+*
+* Arguments: pq_kem_params
+* The following fields of pq_kem_params are used for decapsulate:
+* - unsigned char *shared_secret: pointer to output shared secret
+* (already allocated array of bytes)
+* - unsigned char *cipher_text: pointer to input cipher text
+* (already allocated array of bytes)
+* - const unsigned char *private_key: pointer to input constant private key
+*
+* Returns EVP_PQ_SUCCESS on successful decapsulation,
+* return EVP_PQ_FAILURE otherwise and on error
+**************************************************/
+int EVP_kem_decapsulate(pq_kem_params *kem_params);
+
+
+/* SIKE's key generation
+ * Outputs: private key (SIKE_P434_R3_SECRET_KEY_BYTES = SIKE_P434_R3_MSG_BYTES + SIKE_P434_R3_SECRETKEY_B_BYTES + SIKE_P434_R3_PUBLIC_KEY_BYTES bytes)
+ *          public key (SIKE_P434_R3_PUBLIC_KEY_BYTES bytes) */
+int evp_sike_p434_r3_crypto_kem_keypair(OUT unsigned char *public_key, OUT unsigned char *private_key);
+
+/* SIKE's encapsulation
+ * Input:   public key         (SIKE_P434_R3_PUBLIC_KEY_BYTES bytes)
+ * Outputs: shared secret      (SIKE_P434_R3_SHARED_SECRET_BYTES bytes)
+ *          ciphertext message (SIKE_P434_R3_CIPHERTEXT_BYTES = SIKE_P434_R3_PUBLIC_KEY_BYTES + SIKE_P434_R3_MSG_BYTES bytes) */
+int evp_sike_p434_r3_crypto_kem_enc(OUT unsigned char *ciphertext, OUT unsigned char *shared_secret, IN const unsigned char *public_key);
+
+
+/* SIKE's decapsulation
+ * Input:   private_key         (SIKE_P434_R3_SECRET_KEY_BYTES = SIKE_P434_R3_MSG_BYTES + SIKE_P434_R3_SECRETKEY_B_BYTES + SIKE_P434_R3_PUBLIC_KEY_BYTES bytes)
+ *          ciphertext message  (SIKE_P434_R3_CIPHERTEXT_BYTES = SIKE_P434_R3_PUBLIC_KEY_BYTES + SIKE_P434_R3_MSG_BYTES bytes)
+ * Outputs: shared secret ss    (SIKE_P434_R3_SHARED_SECRET_BYTES bytes) */
+int evp_sike_p434_r3_crypto_kem_dec(OUT unsigned char *shared_secret, IN const unsigned char *ciphertext, IN const unsigned char *private_key);
+
+/* SIKE's private constant time copy function.
+ * Current work around for not being able to find a aws-lc
+ * version of constant time copy function. */
+int sike_copy_or_dont(uint8_t * dest, const uint8_t * src, uint32_t len, bool dont);
```

Comment:
I'm convinced awslc _must_ have a constant time copy function somewhere. We just have to look harder :)

More importantly though, as a temporary workaround, this function is only used internally in the SIKE code in one or two places. It isn't something you want to expose to the rest of the library by defining the function in a header. You can keep it "private"/static to the file(s) where it's used.

Reply:
So I was able to find better one for [compare here](https://github.com/awslabs/aws-lc/blob/main/crypto/mem.c#L222-L232), but within I have mostly [this](https://github.com/awslabs/aws-lc/blob/main/crypto/internal.h#L801-L807) being used for copy but it also uses memcpy instead of bit masking stuff.